runtime/stdlib/builtin.go

```diff
@@ -138,3 +140,113 @@ var LogFunction = NewStandardLibraryFunction(
+			},
+		},
+		ReturnTypeAnnotation:  sema.NewTypeAnnotation(sema.PublicKeyType),
+		RequiredArgumentCount: sema.RequiredArgumentCount(2),
+	},
+
+	func(invocation interpreter.Invocation) trampoline.Trampoline {
+		publicKey := invocation.Arguments[0].(*interpreter.ArrayValue)
+		signAlgo := invocation.Arguments[1].(*interpreter.BuiltinCompositeValue)
+
+		value := interpreter.NewPublicKeyValue(publicKey, signAlgo)
+		return trampoline.Done{Result: value}
+	},
+)
+
+// BuiltinValues
+
+var BuiltinValues = StandardLibraryValues{
+	SignatureAlgorithmValue,
+	HashAlgorithmValue,
+}
+
+var SignatureAlgorithmValue = StandardLibraryValue{
+	Name:  sema.SignatureAlgorithmTypeName,
+	Type:  getEnumType(sema.SignatureAlgorithmType, sema.SignatureAlgorithms),
+	Value: getEnumValue(sema.SignatureAlgorithmType, sema.SignatureAlgorithms),
+	Kind:  common.DeclarationKindEnum,
+}
+
+var HashAlgorithmValue = StandardLibraryValue{
+	Name:  sema.HashAlgorithmTypeName,
+	Type:  getEnumType(sema.HashAlgorithmType, sema.HashAlgorithms),
+	Value: getEnumValue(sema.HashAlgorithmType, sema.HashAlgorithms),
+	Kind:  common.DeclarationKindEnum,
+}
+
+func getEnumType(enumType *sema.BuiltinCompositeType, enumCases []sema.BuiltinEnumCase) *sema.SpecialFunctionType {
+	members := make([]*sema.Member, len(enumCases))
+	for _, algo := range enumCases {
+		members[algo.RawValue()] = sema.NewPublicEnumCaseMember(
+			enumType,
+			algo.Name(),
+			algo.DocString(),
+		)
+	}
+
+	constructorType := &sema.SpecialFunctionType{
+		FunctionType: &sema.FunctionType{
+			Parameters: []*sema.Parameter{
+				{
+					Identifier:     sema.EnumRawValueFieldName,
+					TypeAnnotation: sema.NewTypeAnnotation(enumType.EnumRawType),
+				},
+			},
+			ReturnTypeAnnotation: sema.NewTypeAnnotation(
+				&sema.OptionalType{
+					Type: enumType,
+				},
+			),
+		},
+		Members: sema.GetMembersAsMap(members),
+	}
+
+	return constructorType
+}
+
+func getEnumValue(enumType *sema.BuiltinCompositeType, enumCases []sema.BuiltinEnumCase) (value interpreter.Value) {
+	caseCount := len(enumCases)
+	caseValues := make([]*interpreter.BuiltinCompositeValue, caseCount)
+	constructorMembers := interpreter.NewStringValueOrderedMap()
+
+	for _, enumCase := range enumCases {
+		rawValue := enumCase.RawValue()
+		caseValue := interpreter.NewEnumCaseValue(enumType, rawValue)
+		caseValues[rawValue] = caseValue
+		constructorMembers.Set(enumCase.Name(), caseValue)
+	}
+
+	constructor := interpreter.NewHostFunctionValue(
+		func(invocation interpreter.Invocation) trampoline.Trampoline {
+			rawValueArgument := invocation.Arguments[0].(interpreter.IntegerValue).ToInt()
+			var result interpreter.Value = interpreter.NilValue{}
+
+			if rawValueArgument >= 0 && rawValueArgument < caseCount {
+				caseValue := caseValues[rawValueArgument]
+				result = interpreter.NewSomeValueOwningNonCopying(caseValue)
+			}
+
+			return trampoline.Done{Result: result}
+		},
+	)
```

Comment:
It would be great if we would avoid duplicating the code from `declareEnumConstructor` here. 
Can you please try to refactor this constructor function out and use it here in `getEnumValue` and in `declareEnumConstructor`?

Reply:
It causes a circular dependency, that's why had to duplicate it. 
But I'll see if I can move to a common-place and get rid of the recursion.